### ISSUE TITLE
using `tf.image.*` module for one channel or more than three channel? 

### ISSUE THREAD

**Issue by innat** [tensorflow/tensorflow#51049]
**Basic info**

- TensorFlow 2.4
- Colab / Kaggle Platform 

## Question

In `tf`, we have this module [`tf.image`](https://www.tensorflow.org/api_docs/python/tf/image) to perform various image processing operations. If I'm not mistaken, most of the operations require an RGB image, e.g (`h, w, channel == 3`). However, for a reason, I need them to work on one channel (`h, w, channel == 1`) or more than three-channel (`h, w, channel > 3`); basically for a 3D model (..medical imaging). 

## Description 

Let's say, a 3d data set where each sample contains 100 sequences of frames and so technically for a sample, the shape would be (`h, w, 100`). Now, from these 100 frames 5of a sample, there are 4 modalities, and we pick 5 frames for each of these 4 modalities and end up the data set the shape of (`h, w, 20`). 

Now, from this point, we need to do data augmentation with the data set the shape of (`h, w, 20`). In summary

```
(1st modalities): h, w, 5   <- same augmentations through all 5 channel 
(2nd modalities): h, w, 5   <- same augmentations through all 5 channel 
(3rd modalities): h, w, 5   <- same augmentations through all 5 channel 
(4th modalities): h, w, 5   <- same augmentations through all 5 channel 
```

**How can we apply augmentation on each modality with this shape (`h, w, 5`)?**

---

However, if we take only 1 frame from these 4 modalities, then in order to perform `tf.image.*` operation on these frames, there is a dirty approach. So, if we have some like as follows: 

```
(1st modalities): h, w, 1   <- same augmentations through all channel 
(2nd modalities): h, w, 1   <- same augmentations through all channel 
(3rd modalities): h, w, 1   <- same augmentations through all channel 
(4th modalities): h, w, 1   <- same augmentations through all channel 
```

then we can do something like below, for example: 

```python
# split them first 
split_img <- tf.split(h, w, 4)

aug_img = [] # will contain augmented image 

for i, frame in enumerate(split_img): # iterate each of the frame 
   print(frame.shape) # h, w, 1  

   # repeat 3 times to make fake rgb type. 
   img = tf.repeat(frame, repeats=3, axis=-1)
   print(img.shape) # h, w, 3

   # data augmentaiton with tf.image.* module 
   # seed ensure same augmentation for 
   # particular modality 
   img = tf.image.stateless_random_*(img , seed=(i, 2))
   print(img.shape) # h, w, 3
   
   # split the rgb and go back to previous state 
   img, _, _ = tf.split(img, 3, axis=-1)
   print(img.shape) # h, w, 1
 
   # appending augmented image 
   aug_img.append(img)

# lastly concate all augmented modalites 
img = tf.concat(aug_img, axis=-1)
print(img.shape) # h, w, 4
```

**Comment by innat** [comment]
This also brings (I think) another issue, how to apply **batch processing** using `tf.image*` module efficiently? I've crossed @mrry answer on SO, [here](https://stackoverflow.com/a/38922451), it's increasing the data loading time too much.

**Comment by rmothukuru** [comment]
@innat,
Image Operations with Single Channel should be possible by converting an Image from RGB to Grey Scale and then use that converted Image for the Operations. However, as per my understanding, it can't be used for 3D Images.

**Comment by innat** [comment]
@rmothukuru 
yes, you're right, thanks. 

May I know about the batch processing? Is [this](https://stackoverflow.com/a/38922451/9215780) currently the only optimizxed workaround?

**Comment by rmothukuru** [comment]
@innat,
I think you can use [tf.data](https://www.tensorflow.org/guide/data#batching_tensors_with_padding) for Batching.

**Comment by innat** [comment]
@rmothukuru 
thank you, I'm already using `tf.data` api for batching. But for some reason still, it's performing slow. So, I'm just wondering if I'm missing some features to make it more feasible and fast efficient.

**Comment by innat** [comment]
@GdoongMathew 
Thanks for your suggestion. I've already used keras preprocessing layers. An issue with that is, generally, it can be applied only on the 2D data set. But as we mentioned above, it needs to be done on the 3D data set. I've raised an [issue here](https://github.com/keras-team/keras/issues/15046) regarding the keras proc layers. But the built-in 2d image processing layer in keras can be applied to 3D data set with a bit of modification, [mention here](https://github.com/keras-team/keras/pull/15137#issuecomment-895198025). 

However, for the above query, I'm not looking for the keras solution rather tf only, with the `tf.image.random*` functionalities. IMO, it should be supported batch processing.
